### PR TITLE
Add generic test cases + switch to robust predicates

### PR DESCRIPTION
**Pull request by bluenote10**
This PR mirrors:

- https://github.com/w8r/martinez/pull/111 which has added robust predicates for the point-on-line check (`signed_area`). I'm not fully sure whether using robust predicates in this case is the best solution (because of [this observation](https://github.com/w8r/martinez/issues/114#issuecomment-573302942)), but for now I would say it makes sense to stay close to the JS implementation.
- https://github.com/w8r/martinez/pull/117 which introduces a GeoJSON based test suite that replaces all the manually coded edge cases and allows for easier visualization of the test cases. This step has revealed that some of the test cases were actually a bit broken.

A test case file is a GeoJSON file that contains both the input and the expected output, as explained [here](https://github.com/w8r/martinez/issues/116). Tests only succeed if the implementation produces exactly the specified values. The output data can be regenerated by running with `REGEN=true cargo test` (identically to `REGEN=true npm run test` on JS side). The big benefit is that no code has to be written for testing edge cases, and we can get a simple overview over what we test against (*):

[test_cases.pdf](https://github.com/21re/rust-geo-booleanop/files/4093977/test_cases.pdf)

Differences to the JS PR:
- I had to disable two test cases (`collapsed_edges_removed.geojson` and `disjoint_union_nesting.geojson`) for Rust, but basically because these tests are quite broken anyway (the expected result on JS side isn't even valid GeoJSON), see the PDF attached to [#117](https://github.com/w8r/martinez/pull/117#issuecomment-576434846). These cases have to be revisited later anyway.
- Two of the test cases needed a tiny numeric adjustment after copying the test case from JS. I checked the differences and they are all caused by numerical fluctuations -- I guess that has to be expected, e.g. there could be a tiny difference in operator associativity between the implementations.

Notes on the Rust implementation: I haven't found a way to make the dynamically discovered test cases show up as separate tests on cargo level (I think the available `#[test]`'s must be statically known). This is a bit unfortunate, because now the first `assert_eq` failure will break the loop over test cases. We could use a boolean `any_test_failed`, but then we won't get the nice diffing of `pretty_assertions::assert_eq`.

(*) If desired I can commit the plotting script as well. It is a tiny Python script with usage `plot_test_cases.py <geojson1> <geojson2> ...`.

### CODE REVIEW

File: tests/src/compact_geojson.rs [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+impl WriteIndented for &mut File {
+    fn write_indented<S: AsRef<str>>(self, indent: i32, s: S) {
+        let indent = " ".repeat(indent as usize);
+        self.write(indent.as_bytes()).expect("Failed to write to file.");
+        self.write(s.as_ref().as_bytes()).expect("Failed to write to file.");
+    }
+
+    fn write_indented_with_substring(self, indent: i32, s: &str) {
```
Comment (bluenote10) [inline tests/src/compact_geojson.rs]: Just noticing this is not used => remove.

File: tests/src/helper.rs [hunk 2]
```diff
@@ -67,3 +78,93 @@ pub fn xy<X: Into<f64>, Y: Into<f64>>(x: X, y: Y) -> Coordinate<f64> {
...
+        _ => panic!(format!("Invalid operation: {}", op)),
+    };
+
+    ExpectedResult{
+        result: multi_polygon,
+        op: op,
+    }
+}
+
+pub fn load_generic_test_case(filename: &str, regenerate: bool) {
```
Comment (bluenote10) [inline tests/src/helper.rs]: Function name isn't fully appropriate anymore => rename to `run_generic_test_case` or so.

File: lib/src/boolean/signed_area.rs [hunk 3]
```diff
@@ -1,12 +1,24 @@
...
+        Coord{x: p2.x.to_f64().unwrap(), y: p2.y.to_f64().unwrap()},
+    );
+    if res > 0. {
+        F::from(1.).unwrap()
+    } else if res < 0. {
+        F::from(-1.).unwrap()
+    } else {
+        F::from(0.).unwrap()
+    }
 }
```
Comment (untoldwind) [inline lib/src/boolean/signed_area.rs]: Actually I'd write this part more like this:

```rust
#[inline]
pub fn coordinate_to_robust<F>(p : Coordinate<F>) -> Coord 
where
    F: Float,
{
    Coord{x: p.x.to_f64().unwrap(), y: p.y.to_f64().unwrap()}
}

#[inline]
pub fn signed_area<F>(p0: Coordinate<F>, p1: Coordinate<F>, p2: Coordinate<F>) -> F
where
    F: Float,
{
    let res = orient2d(
        coordinate_to_robust(p0),
        coordinate_to_robust(p1),
        coordinate_to_robust(p2),
    );
    if res > 0f64 {
        F::one()
    } else if res < 0f64 {
        -F::one()
    } else {
        F::zero()
    }
}
```

... but that is more like a personal preference

File: tests/src/compact_geojson.rs (hunk 1 above)
Comment (untoldwind) [reply tests/src/compact_geojson.rs]: I think inside a test-case it is just fine to have a helper like this.
Probably makes debugging much easier if the geojson are formatted all the same.

File: tests/src/helper.rs (hunk 2 above)
Comment (untoldwind) [reply tests/src/helper.rs]: `run_generic_test_case` sounds about right

File: tests/src/compact_geojson.rs (hunk 1 above)
Comment (bluenote10) [reply tests/src/compact_geojson.rs]: The function wasn't even needed for debugging anymore -- I refactored its functionality into the `indent_block` function above, so really just a redundant left over...

File: lib/src/boolean/signed_area.rs (hunk 3 above)
Comment (bluenote10) [reply lib/src/boolean/signed_area.rs]: Looks good to me, I have quickly pushed all the changes...
